hub3/ead/dao.go

```diff
@@ -0,0 +1,622 @@
+		metsStatusErr := fmt.Errorf("unable to retrieve METS %s HTTP status error: %d", cfg.Link, resp.StatusCode)
+		logMETSError(cfg.ArchiveID, cfg.InventoryID, metsStatusErr.Error())
+
+		return metsStatusErr
+	}
+
+	defer resp.Body.Close()
+
+	mets, err := metsParse(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	mets.CmetsHdr.AttrCREATEDATE = ""
+	mets.CmetsHdr.AttrLASTMODDATE = ""
+
+	var buf bytes.Buffer
+	enc := xml.NewEncoder(&buf)
+	enc.Indent("", "\t")
+
+	if err := enc.Encode(mets); err != nil {
+		return err
+	}
+
+	if _, err := os.Stat(cfg.getMetsFilePath()); os.IsNotExist(err) {
+		if mkDirErr := os.MkdirAll(cfg.getDirPath(), os.ModePerm); mkDirErr != nil {
+			return mkDirErr
+		}
+	}
+
+	return ioutil.WriteFile(
+		cfg.getMetsFilePath(),
+		buf.Bytes(),
+		os.ModePerm,
+	)
+}
+
+func (c *DaoClient) DefaultDaoFn(cfg *DaoConfig) error {
+	if !strings.Contains(cfg.Link, "/gaf/api/mets/v1/") {
+		return fmt.Errorf("invalid daolink to GAF: ' %s '", cfg.Link)
+	}
+
+	return c.PublishFindingAid(cfg)
+}
+
+func validateMetsRequest(r *http.Request) (string, string, error) {
+	spec := chi.URLParam(r, "spec")
+	err := ValidateSpec(spec)
+	if err != nil {
+		return "", "", err
+	}
+
+	uuid := chi.URLParam(r, "UUID")
+	if uuid == "" {
+		return "", "", fmt.Errorf("UUID cannot be empty")
+	}
+
+	return spec, uuid, nil
+}
+
+// Index indexes the stored METS files identified by their UUID
+func (c *DaoClient) Index(w http.ResponseWriter, r *http.Request) {
+	spec, uuid, err := validateMetsRequest(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+	}
+
+	cfg, err := c.GetDaoConfig(spec, uuid)
+	if err != nil {
+		if errors.Is(err, ErrFileNotFound) {
+			http.Error(w, "unknown UUID", http.StatusNotFound)
+			return
+		}
+
+		http.Error(w, err.Error(), http.StatusNotFound)
+
+		return
+	}
+
+	if err := c.PublishFindingAid(&cfg); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+func (c *DaoClient) Delete(archiveID, uuid string) error {
+	cfg, err := c.GetDaoConfig(archiveID, uuid)
+	if errors.Is(err, ErrFileNotFound) {
+		return nil
+	}
+
+	cfg.RevisionKey = "1"
+
+	if err := c.dropOrphans(&cfg); err != nil {
+		return err
+	}
+
+	if err := cfg.Delete(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Delete indexes the stored METS files identified by their UUID
+func (c *DaoClient) HandleDelete(w http.ResponseWriter, r *http.Request) {
+	spec, uuid, err := validateMetsRequest(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	deleteErr := c.Delete(spec, uuid)
+	http.Error(w, deleteErr.Error(), http.StatusInternalServerError)
+}
+
+// DownloadConfig is a handler that returns a stored METS XML for an inventory.
+func (c *DaoClient) DownloadConfig(w http.ResponseWriter, r *http.Request) {
+	spec, uuid, err := validateMetsRequest(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
```

Comment:
Is it ok to not return after the error as is done in similar situations such as HandleDelete?

Reply:
Added the return.